File: cpp/src/tier3.cpp
#include "sesame/tier3.hpp"

#include <string>

#include "sesame/crypto.hpp"

namespace sesame::tier3 {

Bytes aad_for_headers(std::string_view version, std::string_view key_id,
                      std::string_view timestamp, std::string_view nonce,
                      std::optional<std::string_view> scope) {
    std::string s;
    s.append(version);
    s.push_back('\n');
    s.append(key_id);
    s.push_back('\n');
    s.append(timestamp);
    s.push_back('\n');
    s.append(nonce);
    if (scope) {
        s.push_back('\n');
        s.append(*scope);
    }
    return Bytes(s.begin(), s.end());
}

Bytes seal(const std::array<std::uint8_t, KEY_LEN>& key,
           const std::array<std::uint8_t, IV_LEN>& iv, const Bytes& aad, const Bytes& plaintext) {
    return crypto::aes256gcm_seal(key, iv, aad.data(), aad.size(), plaintext.data(),
                                  plaintext.size());
}

std::optional<Bytes> open(const std::array<std::uint8_t, KEY_LEN>& key,
                          const std::array<std::uint8_t, IV_LEN>& iv, const Bytes& aad,
                          const Bytes& ct_and_tag) {
    return crypto::aes256gcm_open(key, iv, aad.data(), aad.size(), ct_and_tag.data(),
                                  ct_and_tag.size());
}

}  // namespace sesame::tier3

File: go/tier3.go
package sesame

import (
	"crypto/aes"
	"crypto/cipher"
	"errors"
	"strings"
)

// AES-256-GCM parameters.
const (
	KeyLen = 32
	IVLen  = 12
	TagLen = 16
)

var errBadIV = errors.New("sesame: IV must be 12 bytes")

// AADForHeaders builds the GCM associated data:
// version LF key-id LF timestamp LF nonce [LF scope].
func AADForHeaders(version, keyID, timestamp, nonce string, scope *string) []byte {
	fields := []string{version, keyID, timestamp, nonce}
	if scope != nil {
		fields = append(fields, *scope)
	}
	return []byte(strings.Join(fields, "\n"))
}

// Seal encrypts plaintext with AES-256-GCM, returning ciphertext || tag.
func Seal(key, iv, aad, plaintext []byte) ([]byte, error) {
	gcm, err := newGCM(key)
	if err != nil {
		return nil, err
	}
	if len(iv) != IVLen {
		return nil, errBadIV
	}
	return gcm.Seal(nil, iv, plaintext, aad), nil
}

// Open decrypts ciphertext || tag; returns an error on authentication failure.
func Open(key, iv, aad, ctAndTag []byte) ([]byte, error) {
	gcm, err := newGCM(key)
	if err != nil {
		return nil, err
	}
	if len(iv) != IVLen {
		return nil, errBadIV
	}
	return gcm.Open(nil, iv, ctAndTag, aad)
}

func newGCM(key []byte) (cipher.AEAD, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	return cipher.NewGCM(block)
}
